### PR TITLE
exploring a fuller resolution to h02

### PR DESCRIPTION
**Pull request by wilsoncusack**
Exploring a fuller fix that prevents any reentrancy before state changes, second pass on https://github.com/code-423n4/2022-12-backed-findings/issues/195

### CODE REVIEW

**Comment by wilsoncusack** [comment]
actually meaningfully improved gas 
before 
CleanShot 2023-01-30 at 11 04 38@2x

after

CleanShot 2023-01-30 at 11 04 00@2x

File: src/PaprController.sol [hunk 1]
```diff
@@ -283,45 +283,31 @@ contract PaprController is
         address sendTo,
         ReservoirOracleUnderwriter.OracleInfo calldata oracleInfo
     ) external override {
-        uint256 price = _purchaseNFTAndUpdateVaultIfNeeded(auction, maxPrice, sendTo);
-        --_vaultInfo[auction.nftOwner][auction.auctionAssetContract].auctionCount;
+        // This function is almost the same as NFTEDA._purchaseNFT
+        // however we inline so we can receive payment first
+        // and execute logic before sending NFT
+        uint256 id;
```
Comment (adamgobes) [inline src/PaprController.sol:289]: is it gas beneficial to init these vars before assigning them?

File: test/paprController/PurchaseLiquidationAuctionNFT.t.sol [hunk 2]
```diff
@@ -316,33 +316,6 @@ contract PurchaseLiquidationAuctionNFT is BasePaprControllerTest {
         controller.purchaseLiquidationAuctionNFT(auction, auction.startPrice, purchaser, oracleInfo);
     }
 
-    /// @dev we want to prevent abuse of our clearing remaining debt when there is a shortfall
-    /// but it is the borrowers last NFT
-    function testReentrancyOnLastNFTDoesNotClearDebt() public {
```
Comment (adamgobes) [inline test/paprController/PurchaseLiquidationAuctionNFT.t.sol:321]: why did we have to remove this test? probably still want to test the vulnerability doesn't exist?

**Review by adamgobes — APPROVED** [review]

File: src/PaprController.sol (hunk 1 above)
Comment (wilsoncusack) [reply src/PaprController.sol:289]: no was just a product of copying code from NFTEDA where some of these were already inited by the functions return values

**Comment by wilsoncusack** [comment]
Decided to go with this: we lose sending someone the NFT before funds are required, but we think this is cleaner and the point is maybe moot given most arbers will be swapping papr from uniswap which will give them the funds upfront, anyway

**Pull request merged**